docs/use_cases/scaling_rag_for_production.md

```diff
@@ -0,0 +1,641 @@
+
+```python
+# Extract content
+content_ds = ds.map(extract_main_content)
+content_ds.count()
+
+```
+
+Awesome, this will be our dataset. Ray Datasets are optimized for performance at scale - which will make productionizing our system easier for us. Having to manually adjust code while your application grows can get very costly and is prone to errors.
+
+### Processing the data
+
+The next three processing steps will consist of chunking, embedding and indexing our data source. Chunking is the process of splitting your documents into multiple smaller parts. Not only will this be necessary to make your data meet the LLM’s context length limits, it also helps to keep contexts specific enough to remain relevant. On the other hand, if your chunks are too small, the information retrieved might become too narrow. The exact chunk size will depend on your data, the models used and your use case. We will use a standard value here that has been used in a lot of applications.
+
+Let’s define our text splitting logic first, we will use a standard text splitter from LangChain:
+
+```python
+from functools import partial
+from langchain.text_splitter import RecursiveCharacterTextSplitter
+
+# Defining our text splitting function
+def chunking(document, chunk_size, chunk_overlap):
+    text_splitter = RecursiveCharacterTextSplitter(
+        separators=["\n\n", "\n"],
+        chunk_size=chunk_size,
+        chunk_overlap=chunk_overlap,
+        length_function=len)
+
+    chunks = text_splitter.create_documents(
+        texts=[document["text"]], 
+        metadatas=[{"path": document["path"]}])
+    return [{"text": chunk.page_content, "path": chunk.metadata["path"]} for chunk in chunks]
+```
+
+Again, utilize map() for scalability:
+
+```python
+chunks_ds = content_ds.flat_map(partial(
+    chunking, 
+    chunk_size=512, 
+    chunk_overlap=50))
+print(f"{chunks_ds.count()} chunks")
+```
+
+### Embedding the data
+
+Why are we doing all this again? To make our data retrievable in an efficient way, right. We want relevant answers to our questions. And to find the most relevant text sections for a query, we can use a  pretrained model to create vector embeddings for both our data chunks and the query itself. By measuring the distance between the chunk embeddings and the query embedding, we can identify the most relevant chunks, typically referred to as the 'top-k' chunks. There are various pretrained models suitable for this task. For convenience, we will continue using LangChain:
+
+```python
+from langchain.embeddings import OpenAIEmbeddings
+from langchain.embeddings.huggingface import HuggingFaceEmbeddings
+import numpy as np
+from ray.data import ActorPoolStrategy
+
+def get_embedding_model(embedding_model_name, model_kwargs, encode_kwargs):
+    embedding_model = HuggingFaceEmbeddings(
+            model_name=embedding_model_name,

```

Comment:
where does `embedding_model_name` comes from? Shouldn't it be great to provide some of opinions about what models you like at the time of writing this article?

Reply:
I think that depends a lot on the use case and model selection + evaluation would be an article in itself, but I agree that it would make sense to add a sentence or two explaining my choice. I mentioned the HF leaderboard but didn't elaborate how I picked a model from it.

Reply:
Oh, sorry, looks like I didn't mention the leaderboard. I did now and explained why I chose the gbe-base model.